### ISSUE TITLE
Make TensorFlow an optional dependency

### ISSUE THREAD

**Issue by rbharath** [deepchem/deepchem#1879]
Current TensorFlow is a hard dependency for DeepChem. I think we should swap TensorFlow to a soft dependency like RDKit. The main reason for this is that as repos like [torchchem](https://github.com/deepchem/torchchem) and [jaxchem](https://github.com/deepchem/jaxchem) mature, they will likely use deepchem's core features (datasets, splitters, featurizers, transformers, hyperparameter optimization). Ideally users of these packages wouldn't need to have tensorflow installed in order to be able to make use of DeepChem.

What do folks think? 
CC @peastman @nd-02110114 @vsomnath @seyonechithrananda

**Comment by peastman** [comment]
I thought the plan was that we would split out everything that isn't specific to TensorFlow (datasets, featurizures, transformers, etc.) into the molnet repo, and DeepChem would be left with all the TensorFlow specific parts.

**Comment by rbharath** [comment]
Right now, the code is pretty tightly coupled, so it's hard to do a split like that cleanly. I tried in https://github.com/deepchem/deepchem/pull/1780 / https://github.com/deepchem/moleculenet/pull/4 to split out moleculenet but the set of changes got really gnarly and I backtracked from it. 

I've been doing a bit of thinking about design in the last few weeks and I have a couple of thoughts on possible directions:

- ***Original Plan:*** Move (datasets, featurizers, transformers, etc.) to molnet repo, and keep TensorFlow specific parts in deepchem.
  - ***Pros:*** By our original design, MoleculeNet becomes the shared resource that multiple downstream packages like deepchem/torchchem/jaxchem use.
  - ***Cons:*** This looks like it will be a very gnarly change that will be a ton of work. Plus, lots of people use DeepChem for things like splitters/featurizers already. These people will need to all shift to using molnet instead. Even if we phase in this change gradually, it's a large effort.
-  ***Alternative Plan:*** Keep (datasets, featurizers, transformers, etc.) in DeepChem repo. Make TensorFlow a soft dependency for 2.4 release, and possibly move TensorFlow codebase to a new tensorchem repo in 2.5+. The idea is that the core deepchem repo becomes ML framework agnostic and other repos like jaxchem/torchchem/tensorchem provide models in different frameworks. The molnet repo will only contain actual dataset loaders and not (datasets, featurizers, transformers, etc.) in this framing.
    - ***Pros:*** Less code has to move since most of DeepChem's modules stay where they are. This is also more or less the same as the original molnet plan, with the difference that the framework-agnostic repo is deepchem instead of molnet.
    - ***Cons:*** This will require some effort now to make TF a soft dependency and eventually to factor out TF parts into separate repo. Users will need to get used to using tensorchem/torchchem/etc for models. 

What do you folks think? I think the gnarliness of my first attempted refactoring pushed me a bit away from the original design to this alternate framing, but I'm pretty open to whatever design direction ends up making the most sense.

**Comment by peastman** [comment]
> I think the gnarliness of my first attempted refactoring pushed me a bit away from the original design to this alternate framing

I think what has to be done is identical either way.  You'll end up with one repo containing TF specific features and another repo containing TF independent features.  The only difference is what those repos are called.  The amount of work to disentangle the two will be the same either way.

**Comment by rbharath** [comment]
Good point. I think it would be useful for either case to make TF a soft dependency as a step towards separating the TF specific and TF independent features. If this sounds reasonable to folks, I can start work on a PR

**Comment by peastman** [comment]
Can you outline how that would work?  Use of TensorFlow isn't just localized to a few methods.  In packages like dc.models and dc.rl it's pervasive.  Having to import it in every method or function instead of just once at the start of the file would be really cumbersome.

**Comment by rbharath** [comment]
I'm thinking we could do a top level guard around files that are pure tensorflow. For example, in `deepchem/models/__init__.py` perhaps we could do something like

```
try:
  import tensorflow as tf
  from deepchem.models.keras_model import KerasModel
  # More TF specific imports.
except ImportError:
  logger.warning("TensorFlow not installed so modules that require TensorFlow not loaded.")
```

Does this seem like a reasonable design?

**Comment by peastman** [comment]
Yes, that could work.  There also are a lot of test cases that explicitly depend on TensorFlow, so you'll have to deal with those somehow.
